aggregator/src/aggregator/aggregate_share.rs

```diff
@@ -74,46 +120,170 @@ pub(crate) async fn compute_aggregate_share<
+        BatchAggregationsKey<B::BatchIdentifier>,
+        Cow<'a, BatchAggregation<SEED_SIZE, B, A>>,
+    >,
+    task_id: TaskId,
+    aggregation_param: A::AggregationParam,
+}
+
+impl<'a, const SEED_SIZE: usize, B, A> BatchAggregationsIterator<'a, SEED_SIZE, B, A>
+where
+    B: CollectableBatchMode,
+    A: AsyncAggregator<SEED_SIZE>,
+{
+    pub(crate) fn new<
+        InputIterator: IntoIterator<Item = Cow<'a, BatchAggregation<SEED_SIZE, B, A>>>,
+    >(
+        task: &AggregatorTask,
+        batch_aggregation_shard_count: u64,
+        batch_identifier: &B::BatchIdentifier,
+        aggregation_param: &A::AggregationParam,
+        real_batch_aggregations: InputIterator,
+    ) -> Self {
+        // Construct iterator over all possible batch aggregations for the collection identifier.
+        let expected_batch_aggregations = iproduct!(
+            B::batch_identifiers_for_collection_identifier(task.time_precision(), batch_identifier),
+            0..batch_aggregation_shard_count
+        );
+
+        Self {
+            expected_batch_aggregations,
+            real_batch_aggregations: HashMap::from_iter(real_batch_aggregations.into_iter().map(
+                |ba| {
+                    (
+                        BatchAggregationsKey {
+                            batch_identifier: ba.batch_identifier().clone(),
+                            ord: ba.ord(),
+                        },
+                        ba,
+                    )
+                },
+            )),
+            task_id: *task.id(),
+            aggregation_param: aggregation_param.clone(),
+        }
+    }
+}
+
+impl<'a, const SEED_SIZE: usize, B, A> Iterator for BatchAggregationsIterator<'a, SEED_SIZE, B, A>
+where
+    B: CollectableBatchMode,
+    A: AsyncAggregator<SEED_SIZE>,
+{
+    // The iterator yields a tuple of the BatchAggregation and a boolean indicating whether it's
+    // real or a synthetic, empty BA.
+    type Item = (Cow<'a, BatchAggregation<SEED_SIZE, B, A>>, bool);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // See what the next (batch_identifier, ord) we want to yield is
+        let key = self
+            .expected_batch_aggregations
+            .next()
+            .map(|(batch_identifier, ord)| BatchAggregationsKey {
+                batch_identifier,
+                ord,
+            })?;
+
+        // If we have a real BA for that key, yield it, removing it from the HashMap. This lets us
+        // return a value without cloning, and since this iterator gets consumed, the value can't be
+        // yielded again anyway.
+        self.real_batch_aggregations
+            .remove(&key)
+            .map(|real_ba| (real_ba, true))
+            .or_else(|| {
+                // If there was no real BA, synthesize an empty one. This is why we have to yield an
+                // owned value in the other case: we can't instantiate a BA here and return a
+                // reference.
```

Comment:
nit: This comment should be updated/trimmed down to reflect that we're now using a Cow.